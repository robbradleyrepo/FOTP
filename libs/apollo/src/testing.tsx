import { MockedResponse, MockLink } from "@apollo/react-testing";
import { ApolloClientType, NextPageWithApollo } from "@sss/apollo";
import { Locale } from "@sss/i18n";
import { getMockedRouter } from "@sss/next/testing";
import {
  act,
  render,
  RenderOptions,
  RenderResult,
} from "@testing-library/react";
import {
  InMemoryCache,
  InMemoryCacheConfig,
  IntrospectionFragmentMatcher,
  IntrospectionResultData,
} from "apollo-cache-inmemory";
import { ApolloClient } from "apollo-client";
import {
  GetServerSideProps,
  GetServerSidePropsContext,
  GetStaticProps,
} from "next";
import { RouterContext } from "next/dist/shared/lib/router-context";
import { Router } from "next/router";
import { ParsedUrlQuery, stringify } from "querystring";
import React, { ComponentType } from "react";

import { initializeApollo } from "./client";
jest.mock("./client");

interface BaseSandbox {
  // The global mocks to apply
  globalMocks?: MockedResponse[];
  // Additional mocks under test
  mocks?: MockedResponse[];
  renderOptions?: Omit<RenderOptions, "queries">;
  pathname?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  params?: { [key: string]: any };
}

interface PageSandbox extends BaseSandbox {
  NextPage: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    default: React.ComponentType<any>;
    getStaticProps?: GetStaticProps;
    getServerSideProps?: GetServerSideProps;
  };
}

interface PageResult {
  mockLink: MockLink;
  redirect?: { permanent: boolean; destination: string };
  notFound?: boolean;
  page: RenderResult;
  router: Router;
}

interface ComponentSandbox extends BaseSandbox {
  Component: React.ReactNode;
}

interface ComponentResult {
  mockLink: MockLink;
  component: RenderResult;
  router: Router;
}

interface SandboxInit {
  globalMocks?: MockedResponse[];
  introspectionResultData?: IntrospectionResultData;
  locale: Locale;
  Wrapper: ComponentType<{
    apolloClient: ApolloClientType;
    locale: Locale;
  }>;
}

export const createApolloClient = (
  mocks: MockedResponse[],
  introspectionQueryResultData?: IntrospectionResultData
) => {
  const cacheConfig: InMemoryCacheConfig = { addTypename: true };

  // `InMemoryCache` doesn't like `fragmentMatcher: undefined`. We must omit it.
  if (introspectionQueryResultData) {
    cacheConfig.fragmentMatcher = new IntrospectionFragmentMatcher({
      introspectionQueryResultData,
    });
  }

  return new ApolloClient({
    cache: new InMemoryCache(cacheConfig),
    link: new MockLink(mocks, true),
  });
};

export default class Sandbox {
  public constructor(private opts: SandboxInit) {}

  protected createApolloClient(
    mocks: MockedResponse[] = [],
    globalMocks: MockedResponse[] = this.opts.globalMocks ?? []
  ) {
    return createApolloClient(
      [...globalMocks, ...mocks],
      this.opts.introspectionResultData
    );
  }

  public async page({
    NextPage: NextPageBundle,
    globalMocks,
    mocks,
    renderOptions,
    pathname = "/",
    params = {},
  }: PageSandbox): Promise<PageResult> {
    const router = getMockedRouter(pathname, params);

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (global as any).window = Object.create(window);
    Object.defineProperty(window, "location", {
      value: {
        href: `https://localhost${router.asPath}`,
        pathname: router.asPath,
        search: params ? stringify(params) : "",
      },
      writable: true,
    });

    const apolloClient = this.createApolloClient(mocks, globalMocks);

    (initializeApollo as jest.Mock).mockReturnValueOnce(apolloClient);

    const { default: PageComponent } = NextPageBundle;

    const pageData = NextPageBundle.getStaticProps
      ? await NextPageBundle.getStaticProps({ locale: "en-US", params })
      : NextPageBundle.getServerSideProps
      ? await NextPageBundle.getServerSideProps(
          // `getServerSideProps` context argument also requires
          // `req: IncomingMessage` and `res: ServerResponse`. Cast
          // this for testing purposes.
          ({
            locale: "en-US",
            params,
            req: {},
            res: {
              setHeader: jest.fn(),
              statusCode: null,
            },
          } as unknown) as GetServerSidePropsContext<ParsedUrlQuery>
        )
      : { props: {} };

    let page;

    act(() => {
      const Wrapper = this.opts.Wrapper;
      page = render(
        "props" in pageData &&
          !("notFound" in pageData) &&
          !("redirect" in pageData) ? (
          <RouterContext.Provider value={router}>
            <Wrapper apolloClient={apolloClient} locale={this.opts.locale}>
              <PageComponent {...pageData.props} />
            </Wrapper>
          </RouterContext.Provider>
        ) : (
          <div />
        ),
        {
          ...renderOptions,
        }
      );
    });

    return {
      link: apolloClient.link,
      notFound: "notFound" in pageData && pageData.notFound,
      page,
      redirect: "redirect" in pageData && pageData.redirect,
      router,
    } as any; // eslint-disable-line @typescript-eslint/no-explicit-any
  }

  public async component({
    Component,
    ...options
  }: ComponentSandbox): Promise<ComponentResult> {
    const NextPageComponent: NextPageWithApollo = () => {
      return <>{Component}</>;
    };

    const { page: component, ...rest } = await this.page({
      NextPage: { default: NextPageComponent },
      ...options,
    });

    return {
      component,
      ...rest,
    };
  }
}
