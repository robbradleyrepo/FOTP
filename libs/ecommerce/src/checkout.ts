import {
  Order,
  orderFragment,
  Transaction,
  transactionFragment,
} from "@sss/ecommerce/order";
import {
  ProductCore,
  productCoreFragment,
  Variant,
  variantFragment,
} from "@sss/ecommerce/product";
import gql from "graphql-tag";
import { useRouter } from "next/router";
import { RequireAtLeastOne } from "type-fest";

import { AddressInput, PartialAddressInput } from "./address";
import type { EnhancedCartLineItem } from "./cart";
import * as common from "./common";
import {
  coreLineItemCollectionFragment,
  coreLineItemFragment,
  moneyFragment,
  propertiesFragment,
} from "./common";

export interface Checkout extends common.CoreLineItemCollection {
  acceptsEmailMarketing: boolean;
  availableShippingRates: {
    ready: boolean;
    shippingRates: ShippingRate[] | null;
  };
  customer: {
    analyticsUserId: string;
    id: string;
  } | null;
  completedAt: string | null;
  discount: {
    amount: common.Money;
    applicable: boolean;
    code: string;
    reason: DiscountReasonCode;
    type: DiscountType;
    value: number;
  } | null;
  email: string | null;
  id: string;
  lineItems: LineItem[];
  noteAttributes: common.Property[] | null;
  rOrders: Order[] | null;
  ready: boolean;
  shippingRate: ShippingRate | null;
  taxesIncluded: boolean;
  taxLines: TaxLine[] | null;
  transaction: Transaction | null;
  webUrl: string;
}

export interface CheckoutData {
  rCheckout: Checkout | null;
}

export interface CheckoutWithVariants extends Checkout {
  lineItems: LineItemWithVariant[];
}

export interface CheckoutWithVariantsData {
  rCheckout: CheckoutWithVariants | null;
}

export interface CheckoutChargeInput {
  authorizationToken?: string;
  token: string;
}

export enum DiscountReasonCode {
  EXPIRED = "EXPIRED",
  INVALID = "INVALID",
  NOT_APPLICABLE = "NOT_APPLICABLE",
  ONCE_PER_EMAIL = "ONCE_PER_EMAIL",
  REQUIRES_EMAIL = "REQUIRES_EMAIL",
  SPEND_THRESHOLD_MAX = "SPEND_THRESHOLD_MAX",
  SPEND_THRESHOLD_MIN = "SPEND_THRESHOLD_MIN",
}

export enum DiscountType {
  FIXED_AMOUNT = "FIXED_AMOUNT",
  PERCENTAGE = "PERCENTAGE",
  SHIPPING = "SHIPPING",
}

export interface LineItem extends common.CoreLineItem {
  id: string;
  productType: string;
  vendor: string;
}

export enum LineItemType {
  CONTAINER = "container",
}

export interface LineItemWithVariant extends LineItem {
  product: ProductCore;
  variant: Variant;
}

export type CheckoutBillingAddressInput = { address: AddressInput };
export type CheckoutDiscountInput = { code: string | null };
export type CheckoutEmailInput = {
  acceptsEmailMarketing?: boolean;
  email?: string;
};
export type CheckoutLineItemInput = Pick<
  LineItem,
  "productId" | "properties" | "quantity" | "variantId"
> & {
  frequency: common.Frequency | null;
};
export type CheckoutPhoneInput = { phone: string | null };
export type CheckoutShippingAddressInput =
  | {
      address: AddressInput;
      partial?: false;
    }
  | {
      address: PartialAddressInput;
      partial: true;
    };
export type CheckoutShippingRateInput = { handle: string };

export type CheckoutNoteInput = { note: string };

export type NoteAttributesInput = common.Property[];

export type CheckoutUpdateInput = RequireAtLeastOne<{
  billingAddress: CheckoutBillingAddressInput;
  discount: CheckoutDiscountInput;
  email: CheckoutEmailInput;
  lineItems: CheckoutLineItemInput[];
  phone: CheckoutPhoneInput;
  shippingAddress: CheckoutShippingAddressInput;
  shippingRate: CheckoutShippingRateInput;
  note: CheckoutNoteInput;
  noteAttributes: NoteAttributesInput;
}>;

export interface CheckoutUpdatePayload {
  rCheckout: Checkout | null;
  userErrors: common.UserError[];
}

export interface CheckoutUpdateData {
  payload: CheckoutUpdatePayload;
}

interface CheckoutChargePayload extends CheckoutUpdatePayload {
  authorizationToken: string | null;
}

export interface CheckoutChargeData {
  payload: CheckoutChargePayload;
}

export interface ShippingRate {
  handle: string;
  price: common.Money;
  title: string;
}

export interface TaxLine {
  price: common.Money;
  rate: number;
  title: string;
}

const checkoutLineItemFragment = gql`
  fragment checkoutLineItem on RLineItem {
    ...coreLineItem

    id
    productType
    vendor
  }
`;

const shippingRateFragment = gql`
  fragment shippingRate on RShippingRate {
    handle
    price {
      ...money
    }
    title
  }
`;

const checkoutFragment = gql`
  fragment checkout on RCheckout {
    ...coreLineItemCollection
    acceptsEmailMarketing
    availableShippingRates {
      ready
      shippingRates {
        ...shippingRate
      }
    }
    customer {
      analyticsUserId
      id
    }
    completedAt
    discount {
      amount {
        ...money
      }
      applicable
      code
      reason
      type
      value
    }
    email
    id
    lineItems {
      ...checkoutLineItem
    }
    noteAttributes {
      ...properties
    }
    rOrders {
      ...order
    }
    ready
    shippingRate {
      ...shippingRate
    }
    taxesIncluded
    taxLines {
      price {
        ...money
      }
      rate
      title
    }
    transaction {
      ...transaction
    }
    webUrl
  }

  ${checkoutLineItemFragment}
  ${shippingRateFragment}
  ${coreLineItemCollectionFragment}
  ${coreLineItemFragment}
  ${moneyFragment}
  ${propertiesFragment}
  ${orderFragment}
  ${transactionFragment}
`;

export const CHECKOUT = gql`
  query CHECKOUT($id: ID) {
    rCheckout(id: $id) {
      ...checkout
    }
  }
  ${checkoutFragment}
`;

export const CHECKOUT_CHARGE = gql`
  mutation CHECKOUT_CHARGE($token: String!, $authorizationToken: String) {
    payload: rCheckoutCharge(
      input: {
        authorizationToken: $authorizationToken
        processor: STRIPE
        token: $token
        type: CREDIT_CARD
      }
    ) {
      authorizationToken
      rCheckout {
        ...checkout
      }
      userErrors {
        code
        field
        message
      }
    }
  }
  ${checkoutFragment}
`;

export const CHECKOUT_COMPLETED = gql`
  query CHECKOUT_COMPLETED($id: ID!) {
    rCheckout: rCheckoutCompleted(id: $id) {
      ...checkout
    }
  }
  ${checkoutFragment}
`;

export const CHECKOUT_COMPLETED_CORE = gql`
  query CHECKOUT_COMPLETED_CORE($id: ID!) {
    rCheckout: rCheckoutCompleted(id: $id) {
      id
      completedAt
    }
  }
`;

export const CHECKOUT_UPDATE = gql`
  mutation CHECKOUT_UPDATE(
    $billingAddress: RCheckoutBillingAddressInput
    $discount: RCheckoutDiscountInput
    $email: RCheckoutEmailInput
    $lineItems: [RCheckoutLineItemInput!]
    $phone: RCheckoutPhoneInput
    $shippingAddress: RCheckoutShippingAddressInput
    $shippingRate: RCheckoutShippingRateInput
    $note: RCheckoutNoteInput
    $noteAttributes: [RPropertyInput!]
  ) {
    payload: rCheckoutUpdate(
      input: {
        billingAddress: $billingAddress
        discount: $discount
        email: $email
        lineItems: $lineItems
        phone: $phone
        shippingAddress: $shippingAddress
        shippingRate: $shippingRate
        note: $note
        noteAttributes: $noteAttributes
      }
    ) {
      userErrors {
        code
        message
        field
      }
      rCheckout {
        ...checkout
      }
    }
  }
  ${checkoutFragment}
`;

export const CHECKOUT_WITH_VARIANTS = gql`
  query CHECKOUT($id: ID) {
    rCheckout(id: $id) {
      ...checkout
      lineItems {
        ...checkoutLineItem
        product {
          ...productCore
        }
        variant {
          ...variant
        }
      }
    }
  }
  ${checkoutFragment}
  ${checkoutLineItemFragment}
  ${productCoreFragment}
  ${variantFragment}
`;

export const cartLineitemsToCheckoutInput = (
  cartLineItems: EnhancedCartLineItem[]
): CheckoutLineItemInput[] => {
  const checkoutLineItemInputs: CheckoutLineItemInput[] = [];

  for (const cartLineItem of cartLineItems) {
    checkoutLineItemInputs.push({
      frequency: cartLineItem.frequency,
      productId: cartLineItem.variant.product.id,
      properties: [{ name: "cartLineItemId", value: cartLineItem.id }],
      quantity: cartLineItem.quantity,
      variantId: cartLineItem.variant.id,
    });

    if (cartLineItem.container) {
      checkoutLineItemInputs.push({
        frequency: null, // Containers are never subscription items
        productId: cartLineItem.container.productId,
        properties: [
          { name: "cartLineItemId", value: cartLineItem.id },
          { name: "lineItemType", value: LineItemType.CONTAINER },
        ],
        quantity: 1, // We'll only ever have one container per line item
        variantId: cartLineItem.container.variantId,
      });
    }
  }

  return checkoutLineItemInputs;
};

export const getDefaultShippingRate = (
  rates: ShippingRate[]
): ShippingRate | null =>
  rates.sort(
    (a, b) =>
      common.moneyFns.toFloat(a.price) - common.moneyFns.toFloat(b.price)
  )[0] ?? null;

export const getUpdatedNoteAttributes = (
  checkout: Checkout,
  noteAttributes: Record<string, string>
): common.Property[] =>
  Object.entries({
    ...(checkout.noteAttributes ?? []).reduce(
      (accum, { name, value }) => ({ ...accum, [name]: value }),
      {} as Record<string, string>
    ),
    ...noteAttributes,
  }).map(([name, value]) => ({
    name,
    value,
  }));

export const useRedirectToWebCheckout = () => {
  const router = useRouter();

  return (replace = false) =>
    router[replace ? "replace" : "push"]({
      pathname: "/checkout/[step]",
      query: { step: "information" },
    });
};
