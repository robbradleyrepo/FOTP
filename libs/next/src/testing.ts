import { Router } from "next/router";
import { ParsedUrlQuery } from "querystring";
import { mocked } from "ts-jest/utils";

import { getAsPath } from "./helpers";

export const getMockedRouter = (pathname: string, params?: ParsedUrlQuery) => {
  const asPath = getAsPath(pathname, params);

  return mocked(({
    asPath,
    events: {
      off: jest.fn(),
      on: jest.fn(),
    },
    isReady: true,
    pathname,
    prefetch: jest.fn(async () => undefined),
    push: jest.fn(),
    query: params,
    replace: jest.fn(),
    route: pathname,
  } as unknown) as Router);
};
