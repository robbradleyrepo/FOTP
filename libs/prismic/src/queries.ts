import gql from "graphql-tag";

export const CUSTOMIZATION = gql`
  query CUSTOMIZATION($handle: String!) {
    pCustomization(uid: $handle) {
      body {
        ... on PCustomizationBodyImageCustomization {
          primary {
            image
            target
          }
          type
        }
        ... on PCustomizationBodyRichTextCustomization {
          primary {
            richText
            target
          }
          type
        }
        ... on PCustomizationBodyTitleCustomization {
          primary {
            target
            title
          }
          type
        }
      }
    }
  }
`;
