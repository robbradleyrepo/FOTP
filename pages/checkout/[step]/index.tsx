export {
  default,
  getStaticPaths,
  getStaticProps,
} from "../../../src/pages/checkout";
