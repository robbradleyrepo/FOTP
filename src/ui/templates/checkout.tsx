import { Checkout } from "@sss/ecommerce/checkout";
import { Money } from "@sss/ecommerce/common";
import React, { FC } from "react";

import { gutterX, percentage, s } from "@/common/ui/utils";

import { Grid, Item } from "../base/grid";
import { ToastRack } from "../base/toast";
import { belt, height } from "../modules/checkout/common";
import Footer from "../modules/checkout/footer";
import Header from "../modules/checkout/header";
import Sidebar from "../modules/checkout/sidebar";
import { theme } from "../styles/theme";

interface CheckoutTemplateProps {
  checkout: Checkout;
  shippingThreshold: Money | null;
}

const CheckoutTemplate: FC<CheckoutTemplateProps> = ({
  checkout,
  children,
  shippingThreshold,
}) => {
  const mainColumnWidth = [
    percentage(1),
    null,
    percentage(1 / 2),
    percentage(6 / 11),
  ];

  return (
    <>
      <Header />
      <div css={s(gutterX, { marginTop: [height.summary, null, 0] })}>
        <Grid
          _css={s(belt)}
          gx={[theme.spacing.lg, null, null, theme.spacing.xxl]}
        >
          <Item width={mainColumnWidth}>
            <div
              css={s((t) => ({
                borderBottomColor: t.color.border.light,
                borderBottomStyle: "solid",
                borderBottomWidth: 1,
                height: "100%",
              }))}
            >
              {children}
            </div>
          </Item>
          <Item
            width={[percentage(1), null, percentage(1 / 2), percentage(5 / 11)]}
          >
            <Sidebar
              _css={s({
                height: [null, null, "100%"],
                left: 0,
                position: ["absolute", null, "static"],
                top: height.header,
                width: "100%",
                zIndex: 999,
              })}
              checkout={checkout}
              shippingThreshold={shippingThreshold}
            />
          </Item>
          <Item width={mainColumnWidth}>
            <Footer />
          </Item>
        </Grid>
      </div>
      <ToastRack
        _css={s({
          height: 0,
          position: "fixed",
          right: 0,
          top: height.header,
          zIndex: 99999,
        })}
      />
    </>
  );
};

export default CheckoutTemplate;
