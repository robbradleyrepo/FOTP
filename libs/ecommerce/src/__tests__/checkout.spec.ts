import { createApolloClient } from "@sss/apollo/testing";

import type { EnhancedCartLineItem } from "../cart";
import {
  cartLineitemsToCheckoutInput,
  Checkout,
  CHECKOUT_UPDATE,
  CheckoutUpdateData,
  CheckoutUpdateInput,
  getDefaultShippingRate,
  getUpdatedNoteAttributes,
} from "../checkout";
import { mockedResponses } from "../testing";
import { PRODUCT_ID, VARIANT_ID } from "../testing/queries/checkout";

describe("checkout update mock queries", () => {
  it("should add billing", async () => {
    const client = createApolloClient([
      mockedResponses.checkoutUpdate.addBillingAddress,
    ]);

    const { data } = await client.mutate<
      CheckoutUpdateData,
      CheckoutUpdateInput
    >({
      mutation: CHECKOUT_UPDATE,
      variables: {
        billingAddress: {
          address: {
            address1: "Industrious, 6060 Center Drive",
            address2: "Floor 10",
            city: "Los Angeles",
            company: "FOTP",
            countryCode: "US",
            firstName: "John",
            lastName: "Doe",
            phone: "",
            province: "California",
            zip: "90045-1598",
          },
        },
      },
    });

    expect(data?.payload.rCheckout?.billingAddress).toMatchObject({
      address1: "Industrious, 6060 Center Drive",
      address2: "Floor 10",
      city: "Los Angeles",
      company: "FOTP",
      country: "United States",
      countryCode: "US",
      firstName: "John",
      lastName: "Doe",
      phone: "",
      province: "California",
      zip: "90045-1598",
    });
  });

  it("should add discount", async () => {
    const client = createApolloClient([
      mockedResponses.checkoutUpdate.addDiscount,
    ]);

    const { data } = await client.mutate<
      CheckoutUpdateData,
      CheckoutUpdateInput
    >({
      mutation: CHECKOUT_UPDATE,
      variables: {
        discount: {
          code: "EXTRA10",
        },
      },
    });

    expect(data?.payload.rCheckout?.discount).toMatchObject({
      amount: {
        amount: "3.44",
        currencyCode: "USD",
      },
      applicable: true,
      code: "EXTRA10",
      reason: null,
      type: "FIXED_AMOUNT",
      value: 3.44,
    });
  });

  it("should add email", async () => {
    const client = createApolloClient([
      mockedResponses.checkoutUpdate.addEmail,
    ]);

    const { data } = await client.mutate<
      CheckoutUpdateData,
      CheckoutUpdateInput
    >({
      mutation: CHECKOUT_UPDATE,
      variables: {
        email: {
          acceptsEmailMarketing: true,
          email: "user@example.com",
        },
      },
    });

    expect(data?.payload.rCheckout?.email).toEqual("user@example.com");
    expect(data?.payload.rCheckout?.acceptsEmailMarketing).toEqual(true);
  });

  it("should replace line item", async () => {
    const client = createApolloClient([
      mockedResponses.checkoutUpdate.replaceLineItems,
    ]);

    const { data } = await client.mutate<
      CheckoutUpdateData,
      CheckoutUpdateInput
    >({
      mutation: CHECKOUT_UPDATE,
      variables: {
        lineItems: [
          {
            frequency: {
              chargeDelayDays: null,
              orderIntervalFrequency: 30,
              orderIntervalUnit: "DAY",
            },
            productId: PRODUCT_ID,
            properties: [],
            quantity: 1,
            variantId: VARIANT_ID,
          },
        ],
      },
    });

    expect(data?.payload.rCheckout?.lineItems.length).toEqual(1);
    expect(data?.payload.rCheckout?.lineItems).toMatchSnapshot();
  });

  it("should add shipping", async () => {
    const client = createApolloClient([
      mockedResponses.checkoutUpdate.addShippingAddress,
    ]);

    const { data } = await client.mutate<
      CheckoutUpdateData,
      CheckoutUpdateInput
    >({
      mutation: CHECKOUT_UPDATE,
      variables: {
        shippingAddress: {
          address: {
            address1: "Industrious, 6060 Center Drive",
            address2: "Floor 10",
            city: "Los Angeles",
            company: "FOTP",
            countryCode: "US",
            firstName: "John",
            lastName: "Doe",
            phone: "",
            province: "California",
            zip: "90045-1598",
          },
        },
      },
    });

    expect(data?.payload.rCheckout?.shippingAddress).toMatchObject({
      address1: "Industrious, 6060 Center Drive",
      address2: "Floor 10",
      city: "Los Angeles",
      company: "FOTP",
      country: "United States",
      countryCode: "US",
      firstName: "John",
      lastName: "Doe",
      phone: "",
      province: "California",
      zip: "90045-1598",
    });
    expect(data?.payload.rCheckout?.availableShippingRates).toMatchObject({
      ready: true,
      shippingRates: {
        handle: "rate",
        price: {
          amount: "0",
          currencyCode: "USD",
        },
        title: "Free US Shipping",
      },
    });
  });

  it("should add shipping", async () => {
    const client = createApolloClient([
      mockedResponses.checkoutUpdate.addShippingHandle,
    ]);

    const { data } = await client.mutate<
      CheckoutUpdateData,
      CheckoutUpdateInput
    >({
      mutation: CHECKOUT_UPDATE,
      variables: {
        shippingRate: {
          handle: "rate",
        },
      },
    });

    expect(data?.payload.rCheckout?.shippingRate).toMatchObject({
      handle: "rate",
      price: {
        amount: "0",
        currencyCode: "USD",
      },
      title: "Free US Shipping",
    });
  });

  it("should add note attributes", async () => {
    const client = createApolloClient([
      mockedResponses.checkoutUpdate.addNoteAttributes,
    ]);

    const { data } = await client.mutate<
      CheckoutUpdateData,
      CheckoutUpdateInput
    >({
      mutation: CHECKOUT_UPDATE,
      variables: {
        noteAttributes: [{ name: "ajsAnonymousId", value: "123" }],
      },
    });

    expect(data?.payload.rCheckout?.noteAttributes).toMatchObject([
      { name: "ajsAnonymousId", value: "123" },
    ]);
  });
});

describe("cartLineitemsToCheckoutInput", () => {
  it("should map cart line item values to a valid checkout line item input", () => {
    expect(
      cartLineitemsToCheckoutInput([
        {
          container: null,
          frequency: null,
          id: "cart-line-item-123",
          quantity: 12,
          variant: {
            id: "variant-123",
            product: {
              id: "product-123",
            },
          },
        },
        {
          container: null,
          frequency: {
            chargeDelayDays: null,
            orderIntervalFrequency: 30,
            orderIntervalUnit: "DAY",
          },
          id: "cart-line-item-456",
          quantity: 34,
          variant: {
            id: "variant-456",
            product: {
              id: "product-456",
            },
          },
        },
      ] as EnhancedCartLineItem[])
    ).toMatchSnapshot();
  });

  it("should map cart line items' containers to separate checkout line items", () => {
    expect(
      cartLineitemsToCheckoutInput([
        {
          container: {
            productId: "container-product-123",
            variantId: "container-variant-123",
          },
          frequency: null,
          id: "cart-line-item-123",
          quantity: 12,
          variant: {
            id: "variant-123",
            product: {
              id: "product-123",
            },
          },
        },
        {
          container: {
            productId: "container-product-123",
            variantId: "container-variant-123",
          },
          frequency: {
            chargeDelayDays: null,
            orderIntervalFrequency: 30,
            orderIntervalUnit: "DAY",
          },
          id: "cart-line-item-456",
          quantity: 34,
          variant: {
            id: "variant-456",
            product: {
              id: "product-456",
            },
          },
        },
      ] as EnhancedCartLineItem[])
    ).toMatchSnapshot();
  });
});

describe("getDefaultShippingRate", () => {
  it("should return the cheapest shipping method", () => {
    const expectedDefault = {
      handle: "default",
      price: {
        amount: "0.00",
        currencyCode: "USD",
      },
      title: "Free shipping",
    };
    const rates = [
      {
        handle: "cheap",
        price: {
          amount: "0.01",
          currencyCode: "USD",
        },
        title: "Cheap shipping",
      },
      expectedDefault,
      {
        handle: "expensive",
        price: {
          amount: "1000.00",
          currencyCode: "USD",
        },
        title: "Expensive shipping",
      },
    ];

    expect(getDefaultShippingRate(rates)).toBe(expectedDefault);
  });

  it("should return `null` if there are no shipping methods", () => {
    expect(getDefaultShippingRate([])).toBeNull();
  });
});

describe("getUpdatedNoteAttributes", () => {
  const mockCheckout = {
    noteAttributes: [{ name: "foo", value: "Foo" }],
  } as Checkout;

  it("should return an array of properties that includes the existing notes and provided attributes", () => {
    expect(getUpdatedNoteAttributes(mockCheckout, { bar: "Bar" })).toEqual([
      { name: "foo", value: "Foo" },
      { name: "bar", value: "Bar" },
    ]);
  });

  it("should replace existing entries with new values", () => {
    expect(
      getUpdatedNoteAttributes(mockCheckout, { bar: "Bar", foo: "Baz" })
    ).toEqual([
      { name: "foo", value: "Baz" },
      { name: "bar", value: "Bar" },
    ]);
  });

  it("should note include unwanted properties", () => {
    const mockDirtyCheckout = ({
      noteAttributes: [
        { __typename: "RProperties", name: "foo", value: "Foo" },
      ],
    } as unknown) as Checkout;

    expect(getUpdatedNoteAttributes(mockDirtyCheckout, { bar: "Bar" })).toEqual(
      [
        { name: "foo", value: "Foo" },
        { name: "bar", value: "Bar" },
      ]
    );
  });
});
