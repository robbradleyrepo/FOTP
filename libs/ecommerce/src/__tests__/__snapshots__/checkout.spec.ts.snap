// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cartLineitemsToCheckoutInput should map cart line item values to a valid checkout line item input 1`] = `
Array [
  Object {
    "frequency": null,
    "productId": "product-123",
    "properties": Array [
      Object {
        "name": "cartLineItemId",
        "value": "cart-line-item-123",
      },
    ],
    "quantity": 12,
    "variantId": "variant-123",
  },
  Object {
    "frequency": Object {
      "chargeDelayDays": null,
      "orderIntervalFrequency": 30,
      "orderIntervalUnit": "DAY",
    },
    "productId": "product-456",
    "properties": Array [
      Object {
        "name": "cartLineItemId",
        "value": "cart-line-item-456",
      },
    ],
    "quantity": 34,
    "variantId": "variant-456",
  },
]
`;

exports[`cartLineitemsToCheckoutInput should map cart line items' containers to separate checkout line items 1`] = `
Array [
  Object {
    "frequency": null,
    "productId": "product-123",
    "properties": Array [
      Object {
        "name": "cartLineItemId",
        "value": "cart-line-item-123",
      },
    ],
    "quantity": 12,
    "variantId": "variant-123",
  },
  Object {
    "frequency": null,
    "productId": "container-product-123",
    "properties": Array [
      Object {
        "name": "cartLineItemId",
        "value": "cart-line-item-123",
      },
      Object {
        "name": "lineItemType",
        "value": "container",
      },
    ],
    "quantity": 1,
    "variantId": "container-variant-123",
  },
  Object {
    "frequency": Object {
      "chargeDelayDays": null,
      "orderIntervalFrequency": 30,
      "orderIntervalUnit": "DAY",
    },
    "productId": "product-456",
    "properties": Array [
      Object {
        "name": "cartLineItemId",
        "value": "cart-line-item-456",
      },
    ],
    "quantity": 34,
    "variantId": "variant-456",
  },
  Object {
    "frequency": null,
    "productId": "container-product-123",
    "properties": Array [
      Object {
        "name": "cartLineItemId",
        "value": "cart-line-item-456",
      },
      Object {
        "name": "lineItemType",
        "value": "container",
      },
    ],
    "quantity": 1,
    "variantId": "container-variant-123",
  },
]
`;

exports[`checkout update mock queries should replace line item 1`] = `
Array [
  Object {
    "__typename": "RLineItem",
    "frequency": Object {
      "__typename": "RFrequency",
      "chargeDelayDays": null,
      "orderIntervalFrequency": 30,
      "orderIntervalUnit": "DAY",
    },
    "id": "E9vEeaxExeZQv38FGeT+UAVTr/o=",
    "image": "https://cdn.shopify.com/s/files/1/0266/2825/9918/products/the-one-x1.png?v=1589206978",
    "linePrice": Object {
      "__typename": "MoneyV2",
      "amount": "34.37",
      "currencyCode": "USD",
    },
    "price": Object {
      "__typename": "MoneyV2",
      "amount": "34.37",
      "currencyCode": "USD",
    },
    "productId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzE=",
    "productType": "",
    "properties": Array [],
    "quantity": 1,
    "sku": "FPTO01-PH",
    "title": "The One",
    "variantId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8x",
    "variantTitle": "1 pouch",
    "vendor": "Front Of The Pack",
  },
]
`;
